### PR TITLE
Non-record: 1.1201 BPB - Shared ValueEmbedding (tok_emb reuse, layers 5-10) + Legal TTT

### PR DESCRIPTION
**Pull request by mradassaad**
## Summary

- Reuses tied `tok_emb` for ValueEmbedding instead of training a separate embedding table, with learned projection + per-layer scales
- Freed parameter budget allows expanding VE from 2 layers (9-10) to 6 layers (5-10)
- Built on PR #549 stack (LeakyReLU(0.5)² + Parameter Banking + Legal TTT)
- **3-seed mean: 1.1201 bpb (std 0.0002)**

## Results (8xH100 80GB SXM)

| Seed | Steps | Pre-TTT bpb | Post-TTT bpb | Artifact |
|------|-------|-------------|-------------|----------|
| 1337 | 7,065 | 1.1221 | **1.1199** | 16,024,522 |
| 42 | 7,066 | 1.1224 | **1.1202** | 15,827,158 |
| 2025 | 7,067 | 1.1225 | **1.1203** | 15,839,486 |
| **Mean** | **7,066** | **1.1223** | **1.1201** | |

## Key Change

```python
# Before (PR #549): separate VE embedding, layers 9-10
self.ve_shared = ValueEmbedding(vocab_size, ve_dim=128, model_dim=kv_dim)

# After: reuse tok_emb, layers 5-10
self.ve_shared = ValueEmbedding(vocab_size, ve_dim=128, model_dim=kv_dim, tok_emb=self.tok_emb)
```

tok_emb already encodes rich token identity from full gradient signal. A separate VE table initialized from scratch can't match that in limited training steps.

## Test plan
- [x] 3-seed validation (1337, 42, 2025) — all under 16MB, consistent results
- [x] Legal TTT protocol verified (score-first, inference_mode)
- [x] Logs included for all seeds

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: 1.1201 BPB - Shared ValueEmbedding (tok_emb reuse, layers 5-10) + Legal TTT

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)


## PR #768 — LoRA TTT submission (2026-03-17_LoRA_TTT)

### Commit: 01e3bc8d728fecca82a44a6abd690c871150836b
### Primary file: records/track_10min_16mb/2026-03-17_LoRA_TTT/train_gpt.py

---

### Check 1: N-gram family bug
No evidence of hash lookups, `primes` arrays, `ctx_hash`, `full_key`, or any target-included hash key. The word "primes" appears only once in a comment about optimizer warmup (line 1167). **CLEAN.**

### Check 2: Pre-Quant TTT on val_tokens (illegal multi-epoch AdamW)
The TTT evaluation function `eval_val_ttt_lora` (line 848) processes val tokens in a single forward pass per chunk — no multi-epoch loop over val data. The optimizer (`Adam`, line 803) is reset per document batch (`lora.reset()` + `_reset_ttt_optimizer`, lines 890-891). **CLEAN.**

### Check 3: Legal score-first-per-chunk TTT
Fully compliant. The chunk loop (approx. line 898+) follows strict score-first discipline:

1. **Forward pass** (line 922-928): grad graph kept alive only when `needs_train=True` (i.e., not the final chunk for any active doc).
2. **Score BEFORE adapt** (lines 930-938): `with torch.no_grad():` accumulates BPB loss into `loss_sum`/`byte_sum` accumulators for the current chunk — scoring happens before any weight update.
3. **Train step** (lines 940-946): `mask = [float(ci < num_chunks[b] - 1) ...]` explicitly zeros out the last chunk for each doc (`is_last_chunk` guard). `(per_doc * mask).sum().backward()` + `cur_opt.step()` only updates on non-final chunks.
4. `needs_train` (line 905) skips grad graph entirely when all docs in the batch are on their last chunk.

This matches the reference pattern from PR #1413.

### Check 4: Scored-region SLOT
No per-window `(nll * mask).sum()` optimization over scored positions found. **NOT PRESENT.**

### Summary
This is a clean, legal LoRA-based score-first TTT implementation. The adapter (`BatchedTTTLoRA`) wraps Q/V projections and the LM head with per-batch-element LoRA weights (rank=8 default). The score-first invariant is strictly maintained: BPB accumulation inside `torch.no_grad()` precedes every gradient update, and the final chunk is masked from training via an explicit per-doc mask tensor.

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern: each chunk scored under `torch.no_grad()` before `optimizer.step()`, with `is_last_chunk` guard preventing adaptation on the final scored chunk.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). TTT implementation follows the legal score-first discipline.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
